data/src/main/java/io/homeassistant/companion/android/data/widgets/WidgetRepositoryImpl.kt

```diff
@@ -46,8 +62,22 @@ class WidgetRepositoryImpl @Inject constructor(
         return loadStringPref(PREF_PREFIX_KEY + PREF_KEY_SERVICE + appWidgetId)
     }
 
-    override suspend fun loadServiceData(appWidgetId: Int): String? {
-        return loadStringPref(PREF_PREFIX_KEY + PREF_KEY_SERVICE_DATA + appWidgetId)
+    override suspend fun loadServiceData(appWidgetId: Int): HashMap<String, Any>? {
+        val count = loadLongPref(PREF_PREFIX_KEY + PREF_KEY_SERVICE_DATA_COUNT + appWidgetId)
+
+        // If the amount of data is zero, return a null map
+        count ?: return null
+
+        val serviceDataMap: HashMap<String, Any> = HashMap()
+        for (i in 0 until count) {
```

Comment:
I had considered doing that, but I was worried about a potential issue with sending multiple values with the same keyword, e.g. sending multiple entity IDs (though that's not possible at the moment, I would like it to be in the future).  However, it looks like multiple entity IDs would have to be stored as a string set of their own.

I can implement string set now, but if multiple entity IDs ever gets implemented, it will require some tweaking.

Reply:
Maybe we should just serialize it to a json object and store as a single string?  The data isn't necessarily flat, could be object inside object as well.  And we shouldn't ever have to modify it only pass when calling service.

Reply:
Oh, that's a really good idea.  That will probably be way faster than doing a bunch of `loadStrings`, too.  I'll have to see how to repack the json into a hashmap so that it can get understood by the request and then sent as a...json....

Maybe I'll change the request, too...

Reply:
Should be as simple as that if you are doing it in the data library.  Might need/want to migrate to map over while doing it.
```
Gson gson = new Gson();
Map map = gson.fromJson(jsonString, Map.class);
```

Reply:
Changed in commit b3d9d06